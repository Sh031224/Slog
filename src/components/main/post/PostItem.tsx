/* eslint-disable jsx-a11y/anchor-is-valid */
import Link from "next/link";
import { memo } from "react";
import styled from "styled-components";
import TimeCounting from "time-counting";
import { AiOutlineEye } from "react-icons/ai";
import { GoCommentDiscussion } from "react-icons/go";

import type { Post } from "types/post";
import timeCalc from "lib/timeCalc";
import numberConverter from "lib/numberConverter";

type Props = {
  item: Post;
  lastEl?: (node?: Element) => void;
};

const PostItem: React.FC<Props> = ({ item, lastEl }) => {
  return (
    <PostItemWrapper ref={lastEl}>
      {item.thumbnail && (
        <Link href={`/post/[idx]`} as={`/post/${item.idx}`}>
          <a>
            <PostThumbnail>
              <PostThumbnailImg src={item.thumbnail} alt={item.title} />
            </PostThumbnail>
          </a>
        </Link>
      )}
      <Link href={`/post/[idx]`} as={`/post/${item.idx}`}>
        <a style={{ display: "flex", height: "100%" }}>
          <PostContent>
            <PostContentTitle>
              {item.title}
              <PostContentDescription hasThumbnail={item.thumbnail !== null}>
                {item.description}
              </PostContentDescription>
            </PostContentTitle>
            <PostContentSubinfo title={timeCalc.getTime(item.created_at)}>
              {TimeCounting(item.created_at, { lang: "ko" })}
            </PostContentSubinfo>
          </PostContent>
        </a>
      </Link>
      <PostItemInfo>
        <PostItemInfoContainer>
          <AiOutlineEye />
          <PostItemInfoCount>{numberConverter(item.view)}</PostItemInfoCount>
          <GoCommentDiscussion />
          <PostItemInfoCount>{numberConverter(item.comment_count)}</PostItemInfoCount>
        </PostItemInfoContainer>
      </PostItemInfo>
    </PostItemWrapper>
  );
};

const PostItemWrapper = styled.div`
  width: 19rem;
  box-shadow: ${({ theme }) => theme.colors.sdBlack} 0px 2px 10px 0px;
  display: flex;
  flex-direction: column;
  background: ${({ theme }) => theme.colors.bgWhite};
  border-radius: 5px;
  transition: box-shadow 0.25s ease-in 0s, transform 0.25s ease-in 0s;
  margin: 1rem;
  overflow: hidden;
  position: relative;
  border-radius: 5px;
  ${({ theme }) => theme.device.desktop} {
    width: calc(50% - 2rem);
  }
  ${({ theme }) => theme.device.tablet} {
    width: 100%;
  }
`;

const PostThumbnail = styled.div`
  position: relative;
  padding-top: 52%;
  cursor: pointer;
`;

const PostThumbnailImg = styled.img`
  width: 100%;
  height: 100%;
  position: absolute;
  top: 0;
  left: 0;
`;

const PostContent = styled.div`
  display: flex;
  flex-direction: column;
  padding: 1rem;
  width: 100%;
  position: relative;
  height: 100%;
  cursor: pointer;
`;

const PostContentTitle = styled.h4`
  margin: 0;
  width: 100%;
  color: ${({ theme }) => theme.colors.ftBlack};
  font-size: 1rem;
  font-weight: 600;
  line-height: 1.5;
  display: -webkit-box;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
  overflow: hidden;
  text-overflow: ellipsis;
  ${({ theme }) => theme.device.tablet} {
    white-space: initial;
  }
`;

const PostContentDescription = styled.div<{ hasThumbnail: boolean }>`
  color: ${({ theme }) => theme.colors.ftGray};
  font-size: 0.8rem;
  font-weight: 400;
  line-height: 1.5;
  text-overflow: ellipsis;
  word-break: break-all;
  display: -webkit-box;
  -webkit-line-clamp: 4;
  -webkit-box-orient: vertical;
  overflow: hidden;
  margin-bottom: ${({ hasThumbnail }) => (hasThumbnail ? "1rem" : "2rem")};
  margin-top: ${({ hasThumbnail }) => (hasThumbnail ? "0.3rem" : "0rem")};
`;

const PostContentSubinfo = styled.span`
  display: block;
  color: ${({ theme }) => theme.colors.ftLightGray};
  font-size: 0.8rem;
  line-height: 1.5;
`;

const PostItemInfo = styled.div`
  border-top: ${({ theme }) => theme.colors.bdDarkWhite} 1px solid;
  padding: 0rem 1rem;
`;

const PostItemInfoContainer = styled.div`
  display: flex;
  align-items: center;
  margin: 0.3rem 0;
  font-size: 1rem;
  color: ${({ theme }) => theme.colors.ftGray};
`;

const PostItemInfoCount = styled.span`
  margin-right: 1rem;
  margin-left: 0.2rem;
  font-size: 0.85rem;
  color: ${({ theme }) => theme.colors.ftLittleGray};
`;

export default memo(PostItem);
