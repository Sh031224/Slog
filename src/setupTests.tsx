import { ThemeProvider } from "styled-components";
import { render } from "@testing-library/react";
import { lightTheme } from "style/theme";
import { store } from "stores";
import { Provider } from "react-redux";
import { NotificationContainer } from "react-notifications";

const Providers = ({ children }) => {
  return (
    <Provider store={store}>
      <ThemeProvider theme={lightTheme}>
        <NotificationContainer />
        {children}
      </ThemeProvider>
    </Provider>
  );
};

const customRender = (ui, options = {}) => render(ui, { wrapper: Providers, ...options });

// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom/extend-expect";
import "jest-styled-components";

(window as any).scrollTo = () => {};

export { customRender as render };
