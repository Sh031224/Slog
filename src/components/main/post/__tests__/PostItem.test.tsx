import { render } from "setupTests";
import PostItem, { IPostItemProps } from "../PostItem";

describe("Post Item", () => {
  const setup = (props: Partial<IPostItemProps> = {}) => {
    const initialProps: IPostItemProps = {
      item: {
        idx: 1,
        title: "Testing",
        description: "This is Testing Post",
        content: "content",
        view: 1,
        is_temp: false,
        fk_category_idx: 1,
        thumbnail: "",
        comment_count: 1,
        created_at: new Date("2020-01-01T00:00:00"),
        updated_at: new Date("2020-01-01T00:00:00")
      }
    };
    return render(<PostItem {...initialProps} {...props} />);
  };

  it("renders correctly", () => {
    const { container } = setup();

    expect(container).toMatchSnapshot();
  });

  it("post href", () => {
    const { container } = setup();

    expect(container.querySelector("a")).toHaveAttribute("href", "/post/1");
  });
});
