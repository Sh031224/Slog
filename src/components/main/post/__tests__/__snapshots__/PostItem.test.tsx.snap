// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post Item renders correctly 1`] = `
.c0 {
  width: 19rem;
  box-shadow: rgba(0,0,0,0.096) 0px 2px 10px 0px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background: #ffff;
  border-radius: 5px;
  -webkit-transition: box-shadow 0.25s ease-in 0s,-webkit-transform 0.25s ease-in 0s;
  -webkit-transition: box-shadow 0.25s ease-in 0s,transform 0.25s ease-in 0s;
  transition: box-shadow 0.25s ease-in 0s,transform 0.25s ease-in 0s;
  margin: 1rem;
  overflow: hidden;
  position: relative;
  border-radius: 5px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 1rem;
  width: 100%;
  position: relative;
  height: 100%;
  cursor: pointer;
}

.c2 {
  margin: 0;
  width: 100%;
  color: black;
  font-size: 1rem;
  font-weight: 600;
  line-height: 1.5;
  display: -webkit-box;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c3 {
  color: #777;
  font-size: 0.8rem;
  font-weight: 400;
  line-height: 1.5;
  text-overflow: ellipsis;
  word-break: break-all;
  display: -webkit-box;
  -webkit-line-clamp: 4;
  -webkit-box-orient: vertical;
  overflow: hidden;
  margin-bottom: 1rem;
  margin-top: 0.3rem;
}

.c4 {
  display: block;
  color: #bdbdbd;
  font-size: 0.8rem;
  line-height: 1.5;
}

.c5 {
  border-top: #dfdfdf85 1px solid;
  padding: 0rem 1rem;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 0.3rem 0;
  font-size: 1rem;
  color: #777;
}

.c7 {
  margin-right: 1rem;
  margin-left: 0.2rem;
  font-size: 0.85rem;
  color: #a0a0a0;
}

@media only screen and (max-width:1250px) {
  .c0 {
    width: calc(50% - 2rem);
  }
}

@media only screen and (max-width:767px) {
  .c0 {
    width: 100%;
  }
}

@media only screen and (max-width:767px) {
  .c2 {
    white-space: initial;
  }
}

<div>
  <div
    class="notification-container notification-container-empty"
  >
    <div />
  </div>
  <div
    class="c0"
  >
    
    <a
      href="/post/1"
      style="display: flex; height: 100%;"
    >
      <div
        class="c1"
      >
        <h4
          class="c2"
        >
          Testing
          <div
            class="c3"
          >
            This is Testing Post
          </div>
        </h4>
        <span
          class="c4"
          title="2020년 01월 01일 00시 00분"
        >
          1년 전
        </span>
      </div>
    </a>
    <div
      class="c5"
    >
      <div
        class="c6"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M942.2 486.2C847.4 286.5 704.1 186 512 186c-192.2 0-335.4 100.5-430.2 300.3a60.3 60.3 0 0 0 0 51.5C176.6 737.5 319.9 838 512 838c192.2 0 335.4-100.5 430.2-300.3 7.7-16.2 7.7-35 0-51.5zM512 766c-161.3 0-279.4-81.8-362.7-254C232.6 339.8 350.7 258 512 258c161.3 0 279.4 81.8 362.7 254C791.5 684.2 673.4 766 512 766zm-4-430c-97.2 0-176 78.8-176 176s78.8 176 176 176 176-78.8 176-176-78.8-176-176-176zm0 288c-61.9 0-112-50.1-112-112s50.1-112 112-112 112 50.1 112 112-50.1 112-112 112z"
          />
        </svg>
        <span
          class="c7"
        >
          1
        </span>
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15 1H6c-.55 0-1 .45-1 1v2H1c-.55 0-1 .45-1 1v6c0 .55.45 1 1 1h1v3l3-3h4c.55 0 1-.45 1-1V9h1l3 3V9h1c.55 0 1-.45 1-1V2c0-.55-.45-1-1-1zM9 11H4.5L3 12.5V11H1V5h4v3c0 .55.45 1 1 1h3v2zm6-3h-2v1.5L11.5 8H6V2h9v6z"
            fill-rule="evenodd"
          />
        </svg>
        <span
          class="c7"
        >
          1
        </span>
      </div>
    </div>
  </div>
</div>
`;
